File: 17_模式匹配/回文树/双端回文树/DoubleEndPalindromeAutomaton.java
package template.string;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class DoubleEndPalindromeAutomaton {
    final int minCharacter;
    final int maxCharacter;
    int range;

    Node odd;
    Node even;

    char[] data;
    int zero;
    int frontSize;
    int backSize;
    Node frontBuildLast;
    Node backBuildLast;
    List<Node> all;

    long palindromeSubstringCnt = 0;

    private Node newNode() {
        Node ans = new Node(range);
        all.add(ans);
        return ans;
    }

    public DoubleEndPalindromeAutomaton(int minCharacter, int maxCharacter, int frontAddition, int backAddition) {
        int cap = frontAddition + backAddition;
        all = new ArrayList<>(2 + cap);
        data = new char[cap];
        zero = frontAddition;
        frontSize = zero - 1;
        backSize = zero;

        odd = newNode();
        odd.len = -1;
        odd.depth = 0;

        even = newNode();
        even.fail = odd;
        even.len = 0;
        even.depth = 0;

        all.clear();
        backBuildLast = frontBuildLast = odd;
    }

    public void buildFront(char c) {
        data[frontSize--] = c;

        int index = c - minCharacter;

        Node trace = frontBuildLast;
        while (frontSize + 2 + trace.len >= backSize) {
            trace = trace.fail;
        }

        while (data[frontSize + trace.len + 2] != c) {
            trace = trace.fail;
        }

        if (trace.next[index] != null) {
            frontBuildLast = trace.next[index];
        } else {
            Node now = newNode();
            now.len = trace.len + 2;
            trace.next[index] = now;

            if (now.len == 1) {
                now.fail = even;
            } else {
                trace = trace.fail;
                while (data[frontSize + trace.len + 2] != c) {
                    trace = trace.fail;
                }
                now.fail = trace.next[index];
            }
            now.depth = now.fail.depth + 1;
            frontBuildLast = now;
        }
        if(frontBuildLast.len == backSize - frontSize - 1){
            backBuildLast = frontBuildLast;
        }
        palindromeSubstringCnt += frontBuildLast.depth;
    }

    public void buildBack(char c) {
        data[backSize++] = c;

        int index = c - minCharacter;

        Node trace = backBuildLast;
        while (backSize - 2 - trace.len <= frontSize) {
            trace = trace.fail;
        }

        while (data[backSize - trace.len - 2] != c) {
            trace = trace.fail;
        }

        if (trace.next[index] != null) {
            backBuildLast = trace.next[index];
        } else {
            Node now = newNode();
            now.len = trace.len + 2;
            trace.next[index] = now;

            if (now.len == 1) {
                now.fail = even;
            } else {
                trace = trace.fail;
                while (data[backSize - trace.len - 2] != c) {
                    trace = trace.fail;
                }
                now.fail = trace.next[index];
            }
            now.depth = now.fail.depth + 1;
            backBuildLast = now;
        }
        if(backBuildLast.len == backSize - frontSize - 1){
            frontBuildLast = backBuildLast;
        }
        palindromeSubstringCnt += backBuildLast.depth;
    }

    public void visit(Consumer<Node> consumer) {
        for (int i = all.size() - 1; i >= 0; i--) {
            consumer.accept(all.get(i));
        }
    }

    public long palindromeSubstringCnt() {
        return palindromeSubstringCnt;
    }

    public int distinctPalindromeSubstring() {
        return all.size();
    }

    public static class Node {
        public Node(int range) {
            next = new Node[range];
        }
        public Node[] next;
        public Node fail;
        public int len;
        public int depth;
    }
}

File: 17_模式匹配/回文树/双端回文树/DoubleEndPalindromeAutomaton.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	loj_141()
}

func demo() {
	tree := NewDoubleEndPalindromeAutomaton(100, 100)
	tree.AppendLeft('a')
	tree.AppendLeft('b')
	tree.AppendLeft('a')
	tree.Append('b')

	// abab
	fmt.Println(tree.PalindromeSubstringCount())
	fmt.Println(tree.DistinctPalindromeSubstring())
	tree.Visit(func(node *TreeNode) {
		fmt.Println(node)
	})
}

// https://loj.ac/p/141
// 1：在字符串 s 的末尾添加一个字符串；
// 2：在字符串 s 的前端添加一个字符串的 反序；
// 3：查询字符串 s 的所有非空回文子串的数量。
func loj_141() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var s string
	fmt.Fscan(in, &s)

	var q int32
	fmt.Fscan(in, &q)

	n := int32(len(s))

	const S int32 = 4e5 + 10
	P := NewDoubleEndPalindromeAutomaton(S, S)
	for i := int32(0); i < n; i++ {
		P.Append(int32(s[i]))
	}

	for i := int32(0); i < q; i++ {
		var op int32
		fmt.Fscan(in, &op)
		if op == 1 {
			var c string
			fmt.Fscan(in, &c)
			for _, v := range c {
				P.Append(v)
			}
		} else if op == 2 {
			var c string
			fmt.Fscan(in, &c)
			for _, v := range c {
				P.AppendLeft(v)
			}
		} else {
			fmt.Fprintln(out, P.PalindromeSubstringCount())
		}
	}
}

type TreeNode struct {
	Next  map[int32]*TreeNode
	Fail  *TreeNode
	Len   int32
	Depth int32
}

func (node *TreeNode) String() string {
	return fmt.Sprintf("TreeNode{Len: %d, Depth: %d, Next: %v}", node.Len, node.Depth, node.Next)
}

// 双端回文树.
type DoubleEndPalindromeAutomaton struct {
	odd, even                     *TreeNode
	data                          []int32
	frontSize, backSize           int32
	frontBuildLast, backBuildLast *TreeNode
	nodes                         []*TreeNode
	palindromeSubstringCount      int
}

func NewDoubleEndPalindromeAutomaton(frontAddition, backAddition int32) *DoubleEndPalindromeAutomaton {
	res := &DoubleEndPalindromeAutomaton{}
	cap_ := frontAddition + backAddition
	nodes := make([]*TreeNode, 0, cap_+2)
	data := make([]int32, cap_)
	zero := frontAddition
	frontSize, backSize := zero-1, zero
	odd := res._alloc()
	odd.Len = -1
	even := res._alloc()
	even.Fail = odd
	frontBuildLast, backBuildLast := odd, odd

	res.odd, res.even = odd, even
	res.data = data
	res.frontSize, res.backSize = frontSize, backSize
	res.frontBuildLast, res.backBuildLast = frontBuildLast, backBuildLast
	res.nodes = nodes
	return res
}

func (tree *DoubleEndPalindromeAutomaton) AppendLeft(char int32) {
	tree.data[tree.frontSize] = char
	tree.frontSize--
	trace := tree.frontBuildLast
	for tree.frontSize+2+trace.Len >= tree.backSize {
		trace = trace.Fail
	}
	for tree.data[tree.frontSize+trace.Len+2] != char {
		trace = trace.Fail
	}
	if v, ok := trace.Next[char]; ok {
		tree.frontBuildLast = v
	} else {
		now := tree._alloc()
		now.Len = trace.Len + 2
		trace.Next[char] = now
		if now.Len == 1 {
			now.Fail = tree.even
		} else {
			trace = trace.Fail
			for tree.data[tree.frontSize+trace.Len+2] != char {
				trace = trace.Fail
			}
			now.Fail = trace.Next[char]
		}
		now.Depth = now.Fail.Depth + 1
		tree.frontBuildLast = now
	}

	if tree.frontBuildLast.Len == tree.backSize-tree.frontSize-1 {
		tree.backBuildLast = tree.frontBuildLast
	}
	tree.palindromeSubstringCount += int(tree.frontBuildLast.Depth)
}

func (tree *DoubleEndPalindromeAutomaton) Append(char int32) {
	tree.data[tree.backSize] = char
	tree.backSize++
	trace := tree.backBuildLast
	for tree.backSize-2-trace.Len <= tree.frontSize {
		trace = trace.Fail
	}
	for tree.data[tree.backSize-trace.Len-2] != char {
		trace = trace.Fail
	}
	if v, ok := trace.Next[char]; ok {
		tree.backBuildLast = v
	} else {
		now := tree._alloc()
		now.Len = trace.Len + 2
		trace.Next[char] = now
		if now.Len == 1 {
			now.Fail = tree.even
		} else {
			trace = trace.Fail
			for tree.data[tree.backSize-trace.Len-2] != char {
				trace = trace.Fail
			}
			now.Fail = trace.Next[char]
		}
		now.Depth = now.Fail.Depth + 1
		tree.backBuildLast = now
	}

	if tree.backBuildLast.Len == tree.backSize-tree.frontSize-1 {
		tree.frontBuildLast = tree.backBuildLast
	}
	tree.palindromeSubstringCount += int(tree.backBuildLast.Depth)
}

// 回文子串总数.
func (tree *DoubleEndPalindromeAutomaton) PalindromeSubstringCount() int {
	return tree.palindromeSubstringCount
}

// 本质不同回文子串总数.
func (tree *DoubleEndPalindromeAutomaton) DistinctPalindromeSubstring() int32 {
	return int32(len(tree.nodes))
}

func (tree *DoubleEndPalindromeAutomaton) Visit(consumer func(*TreeNode)) {
	for i := len(tree.nodes) - 1; i >= 0; i-- {
		consumer(tree.nodes[i])
	}
}

func (tree *DoubleEndPalindromeAutomaton) _alloc() *TreeNode {
	res := &TreeNode{Next: make(map[int32]*TreeNode)}
	tree.nodes = append(tree.nodes, res)
	return res
}
